packages/undo-manager/src/index.js

```diff
@@ -0,0 +1,171 @@
+	 */
+	const newChanges = { ...changes1 };
+	Object.entries( changes2 ).forEach( ( [ key, value ] ) => {
+		if ( newChanges[ key ] ) {
+			newChanges[ key ] = { ...newChanges[ key ], to: value.to };
+		} else {
+			newChanges[ key ] = value;
+		}
+	} );
+
+	return newChanges;
+}
+
+/**
+ * Adds history changes for a single item into a record of changes.
+ *
+ * @param {HistoryRecord}  record  The record to merge into.
+ * @param {HistoryChanges} changes The changes to merge.
+ */
+const addHistoryChangesIntoRecord = ( record, changes ) => {
+	const existingChangesIndex = record?.findIndex(
+		( { id: recordIdentifier } ) => {
+			return typeof recordIdentifier === 'string'
+				? recordIdentifier === changes.id
+				: isShallowEqual( recordIdentifier, changes.id );
+		}
+	);
+	const nextRecord = [ ...record ];
+
+	if ( existingChangesIndex !== -1 ) {
+		// If the edit is already in the stack leave the initial "from" value.
+		nextRecord[ existingChangesIndex ] = {
+			id: changes.id,
+			changes: mergeHistoryChanges(
+				nextRecord[ existingChangesIndex ].changes,
+				changes.changes
+			),
+		};
+	} else {
+		nextRecord.push( changes );
+	}
+	return nextRecord;
+};
+
+/**
+ * Creates an undo manager.
+ *
+ * @return {UndoManager} Undo manager.
+ */
+export function createUndoManager() {
+	/**
+	 * @type {HistoryRecord[]}
+	 */
+	let history = [];
+	/**
+	 * @type {HistoryRecord}
+	 */
+	let cachedRecord = [];
+	/**
+	 * @type {number}
+	 */
+	let offset = 0;
+
+	const dropPendingRedos = () => {
+		history = history.slice( 0, offset || undefined );
+		offset = 0;
+	};
+
+	const appendCachedRecordToLatestHistoryRecord = () => {
+		const index = history.length === 0 ? 0 : history.length - 1;
+		let latestRecord = history[ index ] ?? [];
+		cachedRecord.forEach( ( changes ) => {
+			latestRecord = addHistoryChangesIntoRecord( latestRecord, changes );
+		} );
+		cachedRecord = [];
+		history[ index ] = latestRecord;
+	};
+
+	/**
+	 * Checks whether a record is empty.
+	 * A record is considered empty if it the changes keep the same values.
+	 * Also updates to function values are ignored.
+	 *
+	 * @param {HistoryRecord} record
+	 * @return {boolean} Whether the record is empty.
+	 */
+	const isRecordEmpty = ( record ) => {
+		const filteredRecord = record.filter( ( { changes } ) => {
+			return Object.values( changes ).some(
+				( { from, to } ) =>
+					typeof from !== 'function' &&
+					typeof to !== 'function' &&
+					! isShallowEqual( from, to )
+			);
+		} );
+		return ! filteredRecord.length;
+	};
+
+	return {
+		/**
+		 * Record changes into the stiory.
+		 *
+		 * @param {HistoryRecord=} record   A record of changes to record.
+		 * @param {boolean}        isCached Whether to immediately create an undo point or not.
+		 */
+		addRecord( record, isCached = false ) {
+			const isEmpty = ! record || isRecordEmpty( record );
+			if ( isCached ) {
+				if ( isEmpty ) {
+					return;
+				}
+				record.forEach( ( changes ) => {
```

Comment:
If record is an array, shouldn't it be `addRecords`?

Reply:
It's just a name, but a "record" is an array of changes impacting multiple entities that's why it's an array. It's still a "single record of changes".